File: RailgunVB/Core/Preconditions/BotPerms.vb
Imports Discord
Imports Discord.Commands
Imports Microsoft.Extensions.DependencyInjection
Imports RailgunVB.Core.Configuration

Namespace Core.Preconditions
    
    <AttributeUsage(AttributeTargets.Class Or AttributeTargets.Method, AllowMultiple := True)>
    Public NotInheritable Class BotPerms
        Inherits PreconditionAttribute
    
        Private ReadOnly _guildPermission As GuildPermission? = Nothing
        Private ReadOnly _channelPermission As ChannelPermission? = Nothing
        
        Public Sub New(guildPermission As GuildPermission)
             _guildPermission = guildPermission
        End Sub
        
        Public Sub New(channelPermission As ChannelPermission)
            _channelPermission = channelPermission
        End Sub

        Public Overrides Async Function CheckPermissionsAsync(context As ICommandContext, command As CommandInfo, services As IServiceProvider) As Task(Of PreconditionResult)
            Dim config As MasterConfig = services.GetService(Of MasterConfig)
            Dim self As IGuildUser = Await context.Guild.GetCurrentUserAsync()
            
            If _guildPermission.HasValue
                If Not (self.GuildPermissions.Has(_guildPermission) OrElse self.GuildPermissions.Administrator OrElse 
                   config.DiscordConfig.OtherAdmins.Contains(self.Id) OrElse 
                   self.Id = config.DiscordConfig.MasterAdminId)
                    Return Await Task.FromResult(PreconditionResult.FromError(
                        $"I do not have permission to perform this command! {Format.Bold(
                            $"SERVER-PERM-MISSING : {_guildPermission.ToString()}")}"))
                End If
            ElseIf _channelPermission.HasValue
                Dim channelPerms As ChannelPermissions = self.GetPermissions(context.Channel)
                
                If Not (channelPerms.Has(_channelPermission.Value) OrElse self.GuildPermissions.Administrator OrElse 
                        config.DiscordConfig.OtherAdmins.Contains(self.Id) OrElse 
                        self.Id = config.DiscordConfig.MasterAdminId) 
                    Return Await Task.FromResult(PreconditionResult.FromError(
                        $"I do not have permission to perform this command! {Format.Bold(
                            $"CHANNEL-PERM-MISSING : {_channelPermission.ToString()}")}"))
                End If
            End If
            
            Return Await Task.FromResult(PreconditionResult.FromSuccess())
        End Function
        
    End Class
    
End NameSpace

File: Railgun/Core/Preconditions/BotPerms.cs
using System;
using System.Threading.Tasks;
using Discord;
using Discord.Commands;
using Microsoft.Extensions.DependencyInjection;
using Railgun.Core.Configuration;

namespace Railgun.Core.Preconditions
{
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
    public class BotPerms : PreconditionAttribute
    {
        private readonly GuildPermission? _guildPermission = null;
        private readonly ChannelPermission? _channelPermission = null;

        public BotPerms(GuildPermission guildPermission) => _guildPermission = guildPermission;

        public BotPerms(ChannelPermission channelPermission) => _channelPermission = channelPermission;

        public override async Task<PreconditionResult> CheckPermissionsAsync(ICommandContext context, CommandInfo command, IServiceProvider services) {
            var config = services.GetService<MasterConfig>();
            var self = await context.Guild.GetCurrentUserAsync();

            if (_guildPermission.HasValue) {
                if (!(self.GuildPermissions.Has(_guildPermission.Value) || self.GuildPermissions.Administrator))
                    return await Task.FromResult(PreconditionResult.FromError($"I do not have permission to perform this command! {Format.Bold($"SERVER-PERM-MISSING : {_guildPermission.ToString()}")}"));
            } else if (_channelPermission.HasValue) {
                var channelPerms = self.GetPermissions((IGuildChannel)context.Channel);

                if (!(channelPerms.Has(_channelPermission.Value) || self.GuildPermissions.Administrator))
                    return await Task.FromResult(PreconditionResult.FromError($"I do not have permission to perform this command! {Format.Bold($"CHANNEL-PERM-MISSING : {_channelPermission.ToString()}")}"));
            }

            return await Task.FromResult(PreconditionResult.FromSuccess());
        }
    }
}